internal/provider/provider_resource_id_test.go

```diff
@@ -0,0 +1,98 @@
+package provider
```

Comment:
I have separated out the resources into their own packages as it seems like there are details that are specific to the resources that shouldn’t be exposed, for instance, models and unexported funcs (e.g., `upgradePasswordStateV1toV2()`, `generateHash()` in password), but I'm happy to refactor if people would rather we kept the resources with the provider.

In terms of the tests, I must admit I don't really see them as "resource" tests, we're exercising the provider and a resource it exposes (e.g., password, string etc) through the interaction with TF core , so I see them as provider tests. The tests that are specific to resources themselves (e.g., `TestUpgradePasswordStateV0toV2`) still live with the resource code, but there's an argument for removing these sorts of tests as they're essentially testing "private" (unexported) functions which is generally frowned upon.

Another, indirectly related aspect of the refactoring has to do with import cycles. Perhaps one of the reasons a provider and its resources normally reside within the same package (e.g., `package provider`) is to avoid the issue of import cycles when a resource needs to use the API client that is typically configured in the `Configure()` func, for instance:

```go
func (p *provider) Configure(ctx context.Context, req tfsdk.ConfigureProviderRequest, resp *tfsdk.ConfigureProviderResponse) {
    p.client = client.NewClient()
}
```

Access to the `client` is then [usually achieved in the resource by using something like the following](https://learn.hashicorp.com/tutorials/terraform/plugin-framework-create?in=terraform/providers#explore-resource-order) in the `NewResource()` func:

```go
func (r resource) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
    return resource{
        p: *(p.(*provider)),
    }, nil
}
```

This seems a slightly peculiar pattern to me, as we have a `provider` that has `resources` but each resource also contains the provider and accesses "private" (unexported) fields on the provider in order to get the `client` for instance. 

Although a client is not being used by the random provider, I'd be keen to encourage a pattern whereby the client can be obtained from the provider by checking to see if it implements an interface that allows access to the client without exposing any of the underlying "private" (unexported) fields of the provider. Perhaps something along the following lines:

```go
package client

type Client interface {
     Client() API
}

type API interface {
     GetString() string
}

func NewClient() *client {
    return &client{
        msg: "hello",
    }
}

type client struct {
    msg string
}

func (c *client) GetString() string {
     return c.msg
}
```

```go
package provider

func NewProvider() tfsdk.Provider {
    return &provider{}
}

var _ tfsdk.Provider = (*provider)(nil)

type provider struct {
    client client.API
}

func (p *provider) Client() client.API {
    return p.client
}

func (p *provider) Configure(ctx context.Context, req tfsdk.ConfigureProviderRequest, resp *tfsdk.ConfigureProviderResponse) {
    p.client = client.NewClient()
}
```

```go
package resource

func (r *resourceType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
    c, ok := p.(client.Client)
    if !ok {
        panic("not ok")
    }

    return &resource{
        client: c.Client(),
    }, nil
}

var (
_ tfsdk.Resource                = (*resource)(nil)
_ tfsdk.ResourceWithImportState = (*resource)(nil)
)

type resource struct {
    client client.API
}

func (r *resource) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
    fmt.Println(r.client.GetString())
``` 

So this is a very long-winded way of saying that the refactoring also encourages decoupling the provider and resources.

Reply:
👍 I can't argue against incapsulation: I frankly dislike how Go doesn't support controlling visibility more granularly within the same package, so I guess this is the most idiomatic way to go.

In the future I might copy this approach.

Reply:
Hey folks 👋 I'd like to chat about this more, but it is probably easier to do over a higher bandwidth medium.

In Go terms, the only "publicly facing" things in a provider codebase would be anything outside the `internal/` directory. As long as things abide by that, you can export all types, methods, etc. without suggesting any sort of exported API that needs to worry about compatibility. Historically, most small providers are implemented in a "flat" package structure, e.g. everything terraform-plugin-sdk/framework related in `internal/provider`, to make things easier for contributors and avoid overcomplicating things with extra interfaces, etc.

Reply:
That being said, providers typically avoid this by putting the shared client type in another package so both the provider code and data source/resource code can all reference it without an import cycle.

Reply:
@bflad I've gone ahead and refactored to have all of the resources and provider in `package provider` as I think the case for making things easier for contributors is a fair one. If you can take another look and let me know what you think. Thanks.